core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts

```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
 
     return doc;
   }
+
+  /**
+   * Delegates the render to the original _Angular Universal express engine_.
+   *
+   * In case when the config `reuseCurrentRendering` is enabled and **if there is already a pending
+   * render task for the same rendering key**, it doesn't delegate a new render to Angular Universal.
+   * Instead, it waits for the current rendering to complete and then reuse the result for all waiting requests.
```

Comment:
Hmmm... To be precise, We don't invoke the render callbacks when the timeout passes. We wait until the render finishes and only then we invoke the callbacks. 
And in case we perceive the render as hanging (due to maxRenderTime), we forget those callbacks and never invoke them.